### DISCUSSION TITLE
PiVPN seems to ignore presence of nftables and uses iptables instead

### DISCUSSION THREAD

**Discussion by bartgrefte** [pivpn/pivpn#1727] [Bug Reports]
Not entirely sure if this is a bug or by design, but here goes.

When installing PiVPN on a system with Debian 11 that already has a nftables firewall active, the script still seems to focus on iptables, which results in rules for PiVPN in the iptables firewall even though every other firewall rule is in the nftables firewall.

Can the PiVPN-script be modified/updated to include some sort of a detection mechanism so that when there's a nftables firewall active, it won't add rules to iptables?

As alternative, a yes/no question if the script should add the firewall rules (instead of automatically doing that) would be a solution. Though in that case it might be wise to have the script give some info (the used ip-address range / tunnel interface name) so that the end user knows what to enter in the firewall manually.

Otherwise people will end up with 2 firewalls.

**Comment by coolapso** [comment]
nftables not supported. there was some effort to support this but seems to have been dropped due to complexity and effort for doing so.

**Comment by bartgrefte** [comment]
@4s3ti Nftables has been the default firewall in Debian for a couple of years now, can you at least implement an option to skip the firewall rule creation bit so that nftables users can enter the required rules manually? This way you do not have to implement support for nftables and nftables-users wont be forced to use iptables.

Just out of curiosity, which complexity did you ran into, is it because the input/output/forward chains can have other names than input, output and forward?

**Comment by orazioedoardo** [comment]
> can you at least implement an option to skip the firewall rule creation bit so that nftables users can enter the required rules manually?

> Just out of curiosity, which complexity did you ran into, is it because the input/output/forward chains can have other names than input, output and forward?

This is feasible, though I'd like to have nftables support. The issue is reliably and programmatically adding rules to an existing firewall since nftables chains, as you said, can have arbitrary names, and also many input, etc. chains with different priority, however I'm open to tips on this. Might be worth defaulting to nftables when there is no firewall configured, which is likely the case for the majority of PiVPN users.

**Comment by bartgrefte** [comment]
I tried something in bash and came up with this to convert the existing iptables rules to nftables, while keeping in mind the possible different name of the chain:

```
#!/bin/bash

nft_forward_name=$(nft list table ip filter | grep -B1 forward | grep chain | awk '{print $2}')

iptables -S | grep openvpn | while read -r line
do
    nft=$(iptables-translate ${line/$(echo $line | awk '{print $2}')/$nft_forward_name})
    nft=${nft//\\/}
    $nft
done
```

This for IPv4 only, not 100% sure if this works in every situation, but in Debian 11 it does :)

**Comment by bartgrefte** [comment]
Any updates on this?
